File: 킬링캠프/3주 차 - BFS, DFS/challenge/거리두기 확인하기/Java/BFS/Solution.java
import java.util.*;

class Solution {

    private static final int[][] DIR = new int[][]{{-1, 0}, {0, 1}, {0, -1}, {1, 0}};   // 북, 동, 서, 남

    public int[] solution(String[][] places) {
        int[] result = new int[places.length];
        outer: for (int i = 0; i < places.length; i++) {
            char[][] target = createTarget(places[i]);
            for (int row = 0; row < target.length; row++) {
                for (int col = 0; col < target[row].length; col++) {
                    if (target[row][col] != 'P') {
                        continue;
                    }

                    if (!bfs(target, row, col)) {
                        continue outer;
                    }
                }
            }

            result[i] = 1;
        }

        return result;
    }

    private char[][] createTarget(String[] place) {
        char[][] target = new char[5][5];
        for (int i = 0; i < 5; i++) {
            String s = place[i];
            for (int j = 0; j < 5; j++) {
                target[i][j] = s.charAt(j);
            }
        }

        return target;
    }

    private boolean bfs(char[][] places, int startRow, int startCol) {
        boolean[][] visited = new boolean[places.length][places[0].length];
        Deque<Node> dq = new ArrayDeque<>();
        dq.offer(new Node(startRow, startCol, 0));
        visited[startRow][startCol] = true;

        while (!dq.isEmpty()) {
            Node cur = dq.poll();
            for (int i = 0; i < DIR.length; i++) {
                Node next = new Node(cur.row + DIR[i][0], cur.col + DIR[i][1], cur.step + 1);
                if (!isSafe(places, next) || visited[next.row][next.col] || next.step > 2 || (next.step == 1 && places[next.row][next.col] == 'X')) {
                    continue;
                }

                if (places[next.row][next.col] == 'P') {
                    return false;
                }

                dq.offer(next);
                visited[next.row][next.col] = true;
            }
        }

        return true;
    }

    private boolean isSafe(char[][] places, Node node) {
        return node.row >= 0 && node.row < places.length && node.col >= 0 && node.col < places[node.row].length;
    }
}

class Node {
    int row;
    int col;
    int step;

    Node (int row, int col, int step) {
        this.row = row;
        this.col = col;
        this.step = step;
    }
}

File: 킬링캠프/3주 차 - BFS, DFS/challenge/거리두기 확인하기/BFS/solution.py
from collections import deque

DIR = [[-1, 0], [0, 1], [0, -1], [1, 0]]

def solution(places):
    answer = [1 for _ in range(5)]
    for i in range(5):
        if not has_distance(places[i]):
            answer[i] = 0

    return answer

def has_distance(place):
    for row in range(5):
        for col in range(5):
            if place[row][col] != "P":
                continue

            if not check_distance(place, row, col):
                return False
    
    return True

def check_distance(place, start_row, start_col):
    visited = [[False for _ in range(5)] for _ in range(5)]
    dq = deque()
    dq.append([start_row, start_col, 0])
    visited[start_row][start_col] = True

    while dq:
        current_row, current_col, current_step = dq.popleft()

        for dr, dc in DIR:
            next_row = current_row + dr
            next_col = current_col + dc
            next_step = current_step + 1

            if not is_safe(next_row, next_col, visited) or next_step > 2:
                continue

            if place[next_row][next_col] == "P":
                return False
            elif next_step == 1 and place[next_row][next_col] == "O":
                dq.append([next_row, next_col, next_step])
                visited[next_row][next_col]
    
    return True

def is_safe(row, col, visited):
    return row >= 0 and row < 5 and col >= 0 and col < 5 and not visited[row][col]
